Make `test_cancel_term_ignore` unit test more robust

This unit tests fails occasionally in all CIs. We should have a look into it:

```
________________________ test_cancel_term_ignore[local] ________________________

minimal_job = <reframe.core.schedulers.local._LocalJob object at 0x2b4a75e3d780>
scheduler = <class 'reframe.core.schedulers.local.LocalJobScheduler'>
local_only = None

    def test_cancel_term_ignore(minimal_job, scheduler, local_only):
        # This test emulates a descendant process of the spawned job that
        # ignores the SIGTERM signal:
        #
        #   reframe --- local job script --- sleep_deeply.sh --- sleep
        #                                      (TERM IGN)
        #
        #  Since the "local job script" does not ignore SIGTERM, it will be
        #  terminated immediately after we cancel the job. However, the deeply
        #  spawned sleep will ignore it. We need to make sure that our
        #  implementation grants the sleep process a grace period and then
        #  kills it.
        minimal_job.time_limit = '1m'
        prepare_job(minimal_job,
                    command=os.path.join(test_util.TEST_RESOURCES_CHECKS,
                                         'src', 'sleep_deeply.sh'),
                    pre_run=[''],
                    post_run=[''],
                    prepare_cmds=[''])
        minimal_job.submit()
    
        # Stall a bit here to let the the spawned process start and install its
        # signal handler for SIGTERM
        time.sleep(1)
    
        t_grace = time.time()
        minimal_job.cancel()
        time.sleep(0.1)
        minimal_job.wait()
        t_grace = time.time() - t_grace
    
        # Read pid of spawned sleep
        with open(minimal_job.stdout) as fp:
            sleep_pid = int(fp.read())
    
        assert t_grace >= 2 and t_grace < 5
        assert minimal_job.state == 'FAILURE'
        assert minimal_job.signal == signal.SIGKILL
    
        # Verify that the spawned sleep is killed, too, but back off a bit in
        # order to allow the sleep process to wake up and get the signal
        time.sleep(0.1)
>       assert_process_died(sleep_pid)

unittests/test_schedulers.py:574: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

pid = 28671

    def assert_process_died(pid):
        try:
            os.kill(pid, 0)
            if os.getpid() == 1:
                # We are running in a container; so pid is likely a zombie; reap it
                if os.waitpid(pid, os.WNOHANG)[0] == 0:
                    pytest.fail(f'process {pid} is still alive')
            else:
>               pytest.fail(f'process {pid} is still alive')
E               Failed: process 28671 is still alive

unittests/test_schedulers.py:481: Failed
```

I am reopening this since now it throws an error like the following:

```
# Read pid of spawned sleep
        with open(minimal_job.stdout) as fp:
>           sleep_pid = int(fp.read())
E           ValueError: invalid literal for int() with base 10: ''
```

--- a/unittests/test_schedulers.py
+++ b/unittests/test_schedulers.py
@@ -123,6 +123,20 @@ def prepare_job(job, command='hostname',
         )
 
 
+def get_sleep_pid(job, attempts=3):
+    # Try reading the pid of spawned sleep, until a valid value is retrieved
+    for i in range(attempts):
+        try:
+            with open(job.stdout) as fp:
+                sleep_pid = int(fp.read())
+                return sleep_pid
+        except ValueError:
+            time.sleep(1)
+
+    pytest.fail(f'failed to retrieve the spawned sleep process pid after '
+                f'{attempts} attempts')
+
+
 def assert_job_script_sanity(job):
     '''Assert the sanity of the produced script file.'''
     with open(job.script_filename) as fp:
@@ -130,6 +144,20 @@ def assert_job_script_sanity(job):
                              fp.read())
         assert ['echo prepare', 'echo prerun', 'hostname',
                 'echo postrun'] == matches
+
+
+def get_sleep_pid(job, attempts=3):
+    # Try reading the pid of spawned sleep, until a valid value is retrieved
+    for i in range(attempts):
+        try:
+            with open(job.stdout) as fp:
+                sleep_pid = int(fp.read())
+                return sleep_pid
+        except ValueError:
+            time.sleep(1)
+
+    pytest.fail(f'failed to retrieve the spawned sleep process pid after '
+                f'{attempts} attempts')
 
 
 def _expected_sge_directives(job):
@@ -525,19 +553,7 @@ def test_cancel_with_grace(minimal_job, scheduler, local_only):
     # signal handler for SIGTERM
     time.sleep(1)
 
-    # Try reading the pid of spawned sleep, until a valid value is retrieved
-    for i in range(3):
-        try:
-            with open(minimal_job.stdout) as fp:
-                sleep_pid = int(fp.read())
-        except ValueError:
-            time.sleep(1)
-            continue
-        else:
-            break
-    else:
-        pytest.fail('failed to retrieve the spawned sleep process pid')
-
+    sleep_pid = get_sleep_pid(minimal_job)
     t_grace = time.time()
     minimal_job.cancel()
     time.sleep(0.1)
@@ -579,19 +595,7 @@ def test_cancel_term_ignore(minimal_job, scheduler, local_only):
     # signal handler for SIGTERM
     time.sleep(1)
 
-    # Try reading the pid of spawned sleep, until a valid value is retrieved
-    for i in range(3):
-        try:
-            with open(minimal_job.stdout) as fp:
-                sleep_pid = int(fp.read())
-        except ValueError:
-            time.sleep(1)
-            continue
-        else:
-            break
-    else:
-        pytest.fail('failed to retrieve the spawned sleep process pid')
-
+    sleep_pid = get_sleep_pid(minimal_job)
     t_grace = time.time()
     minimal_job.cancel()
     time.sleep(0.1)

--- a/unittests/test_schedulers.py
+++ b/unittests/test_schedulers.py
@@ -147,20 +147,6 @@ def assert_job_script_sanity(job):
                 'echo postrun'] == matches
 
 
-def get_sleep_pid(job, attempts=3):
-    # Try reading the pid of spawned sleep, until a valid value is retrieved
-    for i in range(attempts):
-        try:
-            with open(job.stdout) as fp:
-                sleep_pid = int(fp.read())
-                return sleep_pid
-        except ValueError:
-            time.sleep(1)
-
-    pytest.fail(f'failed to retrieve the spawned sleep process pid after '
-                f'{attempts} attempts')
-
-
 def _expected_sge_directives(job):
     num_nodes = job.num_tasks // job.num_tasks_per_node
     num_cpus_per_node = job.num_cpus_per_task * job.num_tasks_per_node

--- a/unittests/test_schedulers.py
+++ b/unittests/test_schedulers.py
@@ -130,6 +130,7 @@ def get_sleep_pid(job, attempts=3):
             with open(job.stdout) as fp:
                 sleep_pid = int(fp.read())
                 return sleep_pid
+
         except ValueError:
             time.sleep(1)
 

--- a/unittests/test_schedulers.py
+++ b/unittests/test_schedulers.py
@@ -525,15 +525,24 @@ def test_cancel_with_grace(minimal_job, scheduler, local_only):
     # signal handler for SIGTERM
     time.sleep(1)
 
+    # Try reading the pid of spawned sleep, until a valid value is retrieved
+    for i in range(3):
+        try:
+            with open(minimal_job.stdout) as fp:
+                sleep_pid = int(fp.read())
+        except ValueError:
+            time.sleep(1)
+            continue
+        else:
+            break
+    else:
+        pytest.fail('failed to retrieve the spawned sleep process pid')
+
     t_grace = time.time()
     minimal_job.cancel()
     time.sleep(0.1)
     minimal_job.wait()
     t_grace = time.time() - t_grace
-
-    # Read pid of spawned sleep
-    with open(minimal_job.stdout) as fp:
-        sleep_pid = int(fp.read())
 
     assert t_grace >= 2 and t_grace < 5
     assert minimal_job.state == 'FAILURE'
@@ -570,15 +579,24 @@ def test_cancel_term_ignore(minimal_job, scheduler, local_only):
     # signal handler for SIGTERM
     time.sleep(1)
 
+    # Try reading the pid of spawned sleep, until a valid value is retrieved
+    for i in range(3):
+        try:
+            with open(minimal_job.stdout) as fp:
+                sleep_pid = int(fp.read())
+        except ValueError:
+            time.sleep(1)
+            continue
+        else:
+            break
+    else:
+        pytest.fail('failed to retrieve the spawned sleep process pid')
+
     t_grace = time.time()
     minimal_job.cancel()
     time.sleep(0.1)
     minimal_job.wait()
     t_grace = time.time() - t_grace
-
-    # Read pid of spawned sleep
-    with open(minimal_job.stdout) as fp:
-        sleep_pid = int(fp.read())
 
     assert t_grace >= 2 and t_grace < 5
     assert minimal_job.state == 'FAILURE'

--- a/unittests/test_schedulers.py
+++ b/unittests/test_schedulers.py
@@ -123,21 +123,6 @@ def prepare_job(job, command='hostname',
         )
 
 
-def get_sleep_pid(job, attempts=3):
-    # Try reading the pid of spawned sleep, until a valid value is retrieved
-    for i in range(attempts):
-        try:
-            with open(job.stdout) as fp:
-                sleep_pid = int(fp.read())
-                return sleep_pid
-
-        except ValueError:
-            time.sleep(1)
-
-    pytest.fail(f'failed to retrieve the spawned sleep process pid after '
-                f'{attempts} attempts')
-
-
 def assert_job_script_sanity(job):
     '''Assert the sanity of the produced script file.'''
     with open(job.script_filename) as fp:
@@ -515,6 +500,19 @@ def assert_process_died(pid):
         pass
 
 
+def _read_pid(job, attempts=3):
+    # Try reading the pid of spawned sleep, until a valid value is retrieved
+    for _ in range(attempts):
+        try:
+            with open(job.stdout) as fp:
+                return int(fp.read())
+        except ValueError:
+            time.sleep(1)
+
+    pytest.fail(f'failed to retrieve the spawned sleep process pid after '
+                f'{attempts} attempts')
+
+
 def test_cancel_with_grace(minimal_job, scheduler, local_only):
     # This test emulates a spawned process that ignores the SIGTERM signal
     # and also spawns another process:
@@ -540,7 +538,7 @@ def test_cancel_with_grace(minimal_job, scheduler, local_only):
     # signal handler for SIGTERM
     time.sleep(1)
 
-    sleep_pid = get_sleep_pid(minimal_job)
+    sleep_pid = _read_pid(minimal_job)
     t_grace = time.time()
     minimal_job.cancel()
     time.sleep(0.1)
@@ -582,7 +580,7 @@ def test_cancel_term_ignore(minimal_job, scheduler, local_only):
     # signal handler for SIGTERM
     time.sleep(1)
 
-    sleep_pid = get_sleep_pid(minimal_job)
+    sleep_pid = _read_pid(minimal_job)
     t_grace = time.time()
     minimal_job.cancel()
     time.sleep(0.1)

